examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,534 @@
+        extrapolation="periodic",
+    )
+
+
+# %%
+#
+# Again, let us visualize the effect of this feature expansion on some synthetic hour
+# data with a bit of extrapolation beyond hour=23:
+hour_df = pd.DataFrame(
+    np.linspace(0, 26, 1000).reshape(-1, 1),
+    columns=["hour"],
+)
+splines = periodic_spline_transformer(24, n_knots=12).fit_transform(hour_df)
+splines_df = pd.DataFrame(
+    splines,
+    columns=[f"spline_{i}" for i in range(splines.shape[1])],
+)
+pd.concat([hour_df, splines_df], axis="columns").plot(x="hour")
+_ = plt.title("Periodic spline-based encoding for the 'hour' feature")
+
+
+# %%
+# We can now build a predictive pipeline using this alternative periodic
+# feature engineering strategy.
+#
+# For the "hours" columns we use only 12 knots for a period of 24 hours to
+# avoid over-representing this feature compared to months (12 natural knots)
+# and weekday (7 natural knots).
+cyclic_spline_transformer = ColumnTransformer(
+    transformers=[
+        ("categorical", one_hot_encoder, categorical_columns),
+        ("cyclic_month", periodic_spline_transformer(12), ["month"]),
+        ("cyclic_weekday", periodic_spline_transformer(7), ["weekday"]),
+        ("cyclic_hour", periodic_spline_transformer(24, n_knots=12), ["hour"]),
+    ],
+    remainder="passthrough",
+)
+cyclic_spline_linear_pipeline = make_pipeline(
+    cyclic_spline_transformer,
+    RidgeCV(alphas=alphas),
+)
+evaluate(cyclic_spline_linear_pipeline, X, y, cv=ts_cv)
+
+
+# %%
+# Spline features make it possible for the linear model to successfully
+# leverage the periodic time-related features and reduce the error from ~14% to
+# ~10% of the maximum demand.
+#
+# However linear models cannot capture non-linear interactions between features
+# even if the features them-selves are marginally non-linear as is the case
+# with features constructed by `SplineTransformer`.
+#
+# An example of a such a non-linear interaction that we would like to model
+# could be the impact of the rain that might not be the same during the working
+# days and the week-ends and holidays for instance.
+#
+# To capture such interactions we could either use a partial polynomial
+# expansion `PolynomicalFeatures(degree=2, interaction_only=True)` or use the
+# Nystroem method to compute an approximate polynomial kernel expansion.
+#
+# Let us try the latter for computational reasons:
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.kernel_approximation import Nystroem
+
+
+cyclic_spline_poly_pipeline = make_pipeline(
+    cyclic_spline_transformer,
+    MinMaxScaler(),
+    Nystroem(kernel="poly", degree=2, n_components=300, random_state=0),
+    RidgeCV(alphas=alphas),
+)
+evaluate(cyclic_spline_poly_pipeline, X, y, cv=ts_cv)
+
+# %%
+# We observe that this model performance can almost rival the performance of
+# the gradient boosted trees with an average error around 6% of the maximum
+# demand.
```

Comment:
Not asking to add it here, since it's already quite long but how does polynomial features without cyclic_spline_transformer perform?

Also I wonder if having KBinsDiscretizer would produce mostly equivalent results in terms or score, though yes with fewer artifacts in the prediction.

Reply:
Intuitively I think I would agree with you. Let me try on my local copy what you suggest out of curiosity.

Reply:
Polynomial Features or Polynomial kernel approximation on the raw time features does not work any better than the linear model on the raw time features.

Binning on the other hand only slightly worse than spline features but not by much (one or 2 percents than the matching model, with or without poly kernel approx). Here are the plots when binning:

![binned-linear](https://user-images.githubusercontent.com/89061/122581246-236cd800-d057-11eb-80e4-15d149129dd4.png)
![binned+poly](https://user-images.githubusercontent.com/89061/122581249-24056e80-d057-11eb-8723-0b00a008f065.png)

Reply:
Interesting, thanks for checking it!